Use Unix timer_* API instead of setitimer

pr0me opened the pull request:
Currently, to set and reset timeouts in the _TimeoutExecutor_ the `setitimer` FFI is used.
The function is not exposed in `libc::` and is hence linked in via `extern "C"`, 
required structs are defined manually in LibAFL.
 
As these do not seem to account for different layouts/ pointer sizes, the call fails under armv7
and no alarm is being set (also cf. [this](https://sourceware.org/bugzilla/show_bug.cgi?id=16437)).

While working on this issue, I decided to use the newer timer API (`timer_create, timer_settime`),
as `setitimer` is marked [obsolete](https://linux.die.net/man/2/setitimer).

As this requires global access to a timer_id struct, e.g., in order to disarm timers in the crash handler,
a new trait was introduced for _Executor_s, also streamlining the clock reset between Unix and Windows.

Feedback is welcome.

domenukk commented on libafl/src/bolts/os/unix_signals.rs:
```diff
@@ -13,46 +13,76 @@ use nix::errno::{errno, Errno};
 #[cfg(feature = "std")]
 use std::ffi::CString;
 
-/// armv7 `libc` does not feature a `uncontext_t` implementation
```
Usually we have the comments above `cfg`s

pr0me replied:
aye

domenukk commented on libafl/src/executors/inprocess.rs:
```diff
@@ -114,6 +114,16 @@ where
+    OT: ObserversTuple<I, S>,
+{
+    #[inline]
+    fn reset(&self) {}
```
Maybe we want to name this fn `reset_on_crash` then? But fine like this, too

domenukk commented on libafl/src/executors/timeout.rs:
```diff
@@ -136,23 +113,47 @@ impl<E> TimeoutExecutor<E> {
     /// This should usually be used for `InProcess` fuzzing.
     pub fn new(executor: E, exec_tmout: Duration) -> Self {
         let milli_sec = exec_tmout.as_millis();
-        let it_value = Timeval {
```
We can probably remove `Timeval` if we don't use it anymore

pr0me replied:
woops

domenukk commented on libafl/src/executors/timeout.rs:
```diff
@@ -322,11 +296,32 @@ where
+            let disarmed: libc::itimerspec = zeroed();
+            libc::timer_settime(self.timerid, 0, &disarmed as *const _, null_mut());
+        }
+    }
```
This should call through to the wrapped `reset` fn in case another Executor wants to also use this API

domenukk replied:
Still, the recursive call to `self.executor.reset_on_crash()` (or whatever we're going to call it)

domenukk commented on libafl/src/executors/timeout.rs:
```diff
@@ -322,11 +296,32 @@ where
     fn drop(&mut self) {
         unsafe {
-            windows_delete_timer_queue(self.tp_timer);
+            self.reset();
```
I'd leave the original delete here, in case we call through to some other `reset` fn on crash

pr0me replied:
could you elaborate on this?
On second thought, I am not sure anymore what exactly you mean.

domenukk replied:
We should call through to the inner `reset` in the normal reset fn. But this is the destructor so we shouldn't call the inner resets here

tokatoka replied:
I am not sure why I added this Drop in the first place... 😓 
Now I think we don't even need this destructor

domenukk reviewed (approved):
Sweet, thx! Just some minor nits.

pr0me commented:
Thanks!
I'll work the feedback in, fix CI issues and come back with the changes later.

domenukk commented:
https://unix.stackexchange.com/questions/373705/alternative-to-timer-t-in-macos

Looks like the timer API doesn't exist on iOS, probably we need to leave the existing API around for `#cfg[target_vendor = "apple"]`?

tokatoka commented on libafl/src/executors/inprocess.rs:
```diff
@@ -860,7 +869,7 @@ mod windows_exception_handler {
             compiler_fence(Ordering::SeqCst);
 
-            windows_delete_timer_queue(x);
+            data.reset();
```
Hi 👋 
Do you mean executor.reset() here?

pr0me replied:
yea thanks and sorry for such obvious mistakes, I don't have a Windows test setup and was waiting for CI 😅

tokatoka replied:
ci is broken with the updated rust/clippy btw (on main)
Once it is fixed I can check if it builds on win

domenukk commented on libafl/src/executors/timeout.rs:
```diff
@@ -59,42 +60,29 @@ impl Debug for Timeval {
 
 #[repr(C)]
-#[cfg(unix)]
+#[cfg(all(unix, target_vendor = "apple"))]
```
`#[cfg(target_vendor = "apple")]` should suffice, I don't hope they every do a non-unix OS

domenukk commented on libafl/src/executors/timeout.rs:
```diff
@@ -296,11 +358,10 @@ where
             setitimer(ITIMER_REAL, &mut self.itimerval, null_mut());
             let ret = self.executor.run_target(fuzzer, state, mgr, input);
-            unix_remove_timeout();
+            self.reset_on_crash();
```
Wait, if it's called here, then `reset_on_crash` doesn't seem consistent, sorry!
Maybe `post_run`?

domenukk commented on libafl/src/executors/timeout.rs:
```diff
@@ -322,11 +383,45 @@ where
+
 #[cfg(windows)]
 impl<E> Drop for TimeoutExecutor<E> {
     fn drop(&mut self) {
```
As @tokatoka  suggested, this can completely be removed

tokatoka replied:
Yes, now I've found out that this is a bug! We should definitely remove this Drop impl

tokatoka commented:
Could you please merge main into this branch?

tokatoka commented:
ty! I will take a look at your branch when I have time

tokatoka commented:
it builds on windows so seems good.

tokatoka commented:
so basically timer_create is a kinda _modern_ api that supercedes setitimer right?

pr0me commented:
> so basically timer_create is a kinda _modern_ api that supercedes setitimer right?

Yes: ["POSIX.1-2008 marks getitimer() and setitimer() obsolete, recommending the use of the POSIX timers API (timer_gettime(2), timer_settime(2), etc.) instead."](https://linux.die.net/man/2/setitimer)
Apparently setitimer has accuracy issues. Also, there is no way to distinguish who created the signal in the old API.

pr0me commented:
Ok, so in order not to be required to add the HasPostRunReset trait all over the place, when we want to call it recursively on the
inner Executor, I removed the trait and added the post_run_reset as a trait function to `Executor` instead.

Sorry for the inconvenience.

tokatoka commented:
Aside from clippy errors, we get this.
`/home/runner/work/LibAFL/LibAFL/libafl/src/executors/timeout.rs:349: undefined reference to `timer_settime'`

A quick google search suggests me that we can put this flag: `-lrt` when we link.
So maybe we can add the flag to this file? https://github.com/AFLplusplus/LibAFL/blob/main/fuzzers/libfuzzer_libpng_ctx/src/bin/libafl_cc.rs

pr0me commented:
Okay, finally looks good I guess.

Thanks for all the feedback :)

domenukk commented:
@tokatoka want to give this a try on windows before we merge?

domenukk commented:
At some point we should probably do speed tests, too, it might be that one of the APIs is actually faster than the other(?)

tokatoka commented:
It is crashing, let me see why

tokatoka commented on libafl/src/executors/timeout.rs:
```diff
@@ -277,13 +319,54 @@ where
 
             write_volatile(&mut data.timeout_input_ptr, core::ptr::null_mut());
 
-            self.windows_reset_timeout()?;
```
previously I made two functions for windows timeouts
`windows_reset_timeout()` and `windows_delete_timer_queue()`.
(They are pretty confusing, sorry about that)
They two do different things. 
`reset_timeout()` is about setting the timer to 0 sec
`delete_timer_queue()` is about wiping the timer object off.

So in post_run_reset we should do the same as reset_timeout() because we just want to reset it to zero

but in crash_handler we should do what delete_timer_queue() does because we want to wipe the entire timer object

tokatoka commented:
In a nutshell, 
in `post_run_reset()`, we should call `SetThreadpoolTimer` instead of `CloseThreadpoolTimer` to reset the timer
but here in crash_handler
```
            let executor = (data.executor_ptr as *mut E).as_mut().unwrap();
            // reset timer
            if !data.tp_timer.is_null() {
                executor.post_run_reset();
                data.tp_timer = ptr::null_mut();
            }
```
We should not use executor.post_run_reset() but something that calls `CloseThreadpoolTimer`

tokatoka commented:
you can leave it to me if you can merge this to LibAFL's branch.
currently I guess I can't edit your branch

domenukk commented:
You can [checkout a pr](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally#modifying-an-inactive-pull-request-locally)

But in any case, shouldn't the OS take care of the cleanup after a process dies? 

If not, we should probably add a `exiting` flag to `post_run_reset()` for flexibility

tokatoka commented:
> But in any case, shouldn't the OS take care of the cleanup after a process dies?

mhm..., I think you are right. Yes windows should just clean up these things.
well..., then we can keep our current crash handler changes.

pr0me commented:
bump

@tokatoka if you could be so kind to check out the PR and make the Windows changes you deem necessary, that would be much appreciated :)
If not, no problem, then I'll see what I can do.

tokatoka commented:
#510 should do the job.
now we can merge this 🎉

tokatoka commented:
I have a weird error on #510 and I think that's because I checkout-ed a PR. 😢 
can you apply this change of mine https://github.com/AFLplusplus/LibAFL/pull/510/commits/70130adc79c16d5cfdbc49669725ce25d6d1f7f8
to your branch? @pr0me 
and then we can merge this branch after resolving the conflict.

domenukk commented:
Closing this as all commits were included in #510 , thanks!
